external/urlfrontier/src/main/java/com/digitalpebble/stormcrawler/urlfrontier/StatusUpdaterBolt.java

```diff
@@ -58,134 +62,190 @@ public class StatusUpdaterBolt extends AbstractStatusUpdaterBolt
+    // fairness.
+    private final ReentrantReadWriteLock waitAckLock = new ReentrantReadWriteLock(true);
 
-    private AtomicInteger messagesinFlight = new AtomicInteger();
+    private int maxMessagesInFlight = 100000;
+    private long throttleTimeMS = 10;
+
+    // Faster ways of locking until n messages are processed
+    private Semaphore inFlightSemaphore;
 
     private MultiCountMetric eventCounter;
 
     /** Globally set crawlID * */
-    private String globalCrawlID = null;
-
-    public StatusUpdaterBolt() {
-        waitAck =
-                Caffeine.newBuilder()
-                        .expireAfterWrite(60, TimeUnit.SECONDS)
-                        .removalListener(this)
-                        .build();
-    }
+    private String globalCrawlID;
 
     @Override
     public void prepare(
             Map<String, Object> stormConf, TopologyContext context, OutputCollector collector) {
         super.prepare(stormConf, context, collector);
 
-        // host and port of URL Frontier(s)
-        List<String> addresses = ConfUtils.loadListFromConf("urlfrontier.address", stormConf);
-
-        String address = null;
-
-        if (addresses.size() > 1) {
-            int totalTasks = context.getComponentTasks(context.getThisComponentId()).size();
-            // check that the number of tasks is a multiple of the frontier nodes
-            if (totalTasks < addresses.size()) {
-                String message =
-                        "Needs at least one task per frontier node. "
-                                + totalTasks
-                                + " vs "
-                                + addresses.size();
-                LOG.error(message);
-                throw new RuntimeException();
-            }
-            if (totalTasks % addresses.size() != 0) {
-                String message =
-                        "Number of tasks not a multiple of the number of frontier nodes. "
-                                + totalTasks
-                                + " vs "
-                                + addresses.size();
-                LOG.error(message);
-                throw new RuntimeException();
-            }
-
-            int nodeIndex = context.getThisTaskIndex();
-            int assignment = nodeIndex % addresses.size();
-            Collections.sort(addresses);
-            address = addresses.get(assignment);
-        }
+        long expireAfterNMillisec =
+                ConfUtils.getLong(stormConf, URLFRONTIER_CACHE_EXPIREAFTER_MS_KEY, 60_000L);
 
-        if (address == null) {
-            String host = ConfUtils.getString(stormConf, "urlfrontier.host", "localhost");
-            int port = ConfUtils.getInt(stormConf, "urlfrontier.port", 7071);
-            address = host + ":" + port;
-        }
+        waitAck =
+                Caffeine.newBuilder()
+                        .expireAfterWrite(expireAfterNMillisec, TimeUnit.MILLISECONDS)
```

Comment:
wouldn't a duration in second be easier to read? We don't need this level of precision here.